src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,488 @@
+        ret.push_back(TupleAwareAdd(tuple->fields[i], increment));
+      } else {
+        ret.push_back(tuple->fields[i]);
+      }
+    }
+    return Tuple(ret);
+  }
+
+  void BindAndEmit(Var v, Expr e) {
+    if (adjoint_expr_to_var_.count(e)) {
+      e = adjoint_expr_to_var_[e];
+    } else {
+      if (!e.as<TupleNode>()) {
+        adjoint_expr_to_var_.Set(e, v);
+      }
+    }
+    if (v->IsInstance<DataflowVarNode>()) {
+      builder_->Emit(VarBinding(v, e));
+    } else {
+      builder_->EmitOutput(VarBinding(v, e));
+    }
+  }
+
+  // Init the gradient of the target_var_ and update it in adjoint_expr_map_.
+  void InitGradAsOnes(const Var& var) {
+    Expr ones = MakeOnes(var->shape(), Downcast<DynTensorType>(var->checked_type())->dtype);
+    adjoint_expr_map_.Set(var, ones);
+  }
+
+  // Handle the return value of the AD function.
+  // returns the new return value, which would be like:
+  // Tuple(original_return_value,
+  //       Tuple(adjoint_of_require_grads_1, adjoint_of_require_grads_2, ...))
+  Expr Epilogue(const Array<Var>& require_grads) {
+    // create adjoint variables for inputs, and then bind adjoints
+    Array<Expr> out_adjoints;
+    for (Var var : require_grads) {
+      Var adjoint_var = CreateAdjointVar(var, /*is_datalfow_var=*/false);
+
+      if (adjoint_expr_map_.count(var)) {
+        BindAndEmit(adjoint_var, adjoint_expr_map_[var]);
+      } else {
+        BindAndEmit(adjoint_var,
+                    MakeZeros(var->shape(), Downcast<DynTensorType>(var->checked_type())->dtype));
+      }
+      out_adjoints.push_back(adjoint_var);
+    }
+
+    return Tuple(Array<Expr>{target_var_, Tuple(out_adjoints)});
+  }
+
+  // the block builder of the corresponding GradientMutator, to emit bindings
+  BlockBuilder builder_;
+
+  // the differentiation target
+  Var target_var_;
+  // the return value of the differentiated function
+  Expr return_expr_;
+
+  // var to its adjoints var
+  Map<Var, Var> adjoint_var_map_;
+  // var to its adjoint expr
+  Map<Var, Expr> adjoint_expr_map_;
+  // trace binding
+  Map<Expr, Var> adjoint_expr_to_var_;
+};
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(const IRModule& mod, const GlobalVar& gvar,
+                           const Array<Var>& require_grads)
+      : ExprMutator(mod),
+        mod_(mod),
+        gvar_(std::move(gvar)),
+        require_grads_(std::move(require_grads)) {}
+
+  IRModule Transform() {
+    Function old_func = Downcast<Function>(mod_->Lookup(gvar_));
+    CheckRequireGrads(require_grads_, old_func->params, gvar_->name_hint);
+
+    Function new_func = CopyWithNewParams(old_func);
+    // map the parameter list into new params
+    for (size_t i = 0; i < require_grads_.size(); ++i) {
+      int idx = std::find(old_func->params.begin(), old_func->params.end(), require_grads_[i]) -
+                old_func->params.begin();
+      require_grads_.Set(i, new_func->params[idx]);
+    }
```

Comment:
We have to replace vars in require_grads with the new params created in CopyWithNewParams.

We have two solutions:
- This one. Use index to find the new var
- Let CopyWithNewParams return a map, mapping old vars to new vars. Then call `require_grads_.Map(var_map)`

Looking for some advice on this.

Reply:
Will it be good if we first consider the map (old vars to new vars) by scanning the params list once, and use this map to update require_grads_? This is `O(n log n)` while the current one is `O(n^2)`. Speeds do not differ much in practice but just ask which one do you prefer?

Reply:
That is also a solution. But I think it would be also a little complex?

For the time complexity issue, I thought that most functions don't have too much parameters, so it seems fine